src/object/obj_tx.c

```diff
@@ -1993,23 +2034,62 @@ dc_tx_restart(tse_task_t *task)
 {
 	daos_tx_restart_t	*args;
 	struct dc_tx		*tx;
+	uint32_t		 backoff = 0;
 	int			 rc = 0;
 
 	args = dc_task_get_args(task);
 	D_ASSERTF(args != NULL,
 		  "Task Argument OPC does not match DC OPC (restart)\n");
 
-	tx = dc_tx_hdl2ptr(args->th);
+	tx = tse_task_get_priv(task);
 	if (tx == NULL) {
-		rc = -DER_NO_HDL;
-	} else {
-		rc = dc_tx_restart_internal(tx);
+		/* Executing task for the first time. */
 
-		/* -1 for hdl2ptr */
+		tx = dc_tx_hdl2ptr(args->th);
+		if (tx == NULL) {
+			rc = -DER_NO_HDL;
+			goto out;
+		}
+
+		D_MUTEX_LOCK(&tx->tx_lock);
+
+		rc = dc_tx_restart_begin(tx, &backoff);
+		if (rc != 0)
+			goto out_tx_lock;
+
+		if (backoff == 0) {
+			dc_tx_restart_end(tx);
+		} else {
+			/*
+			 * Reinitialize task with a delay to implement the
+			 * backoff and call dc_tx_restart_end.
+			 */
+			rc = tse_task_reinit_with_delay(task, backoff);
+			if (rc != 0) {
+				/* Skip the backoff. */
+				backoff = 0;
+				dc_tx_restart_end(tx);
+				goto out_tx_lock;
+			}
+			D_MUTEX_UNLOCK(&tx->tx_lock);
+			/* Pass our tx reference to task. */
+			tse_task_set_priv(task, tx);
```

Comment:
Where the reference for the "tx" will be released?

Reply:
Good catch; this is a leak. Should be released by the "re-executing task" branch below.